File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.cpp
/**
 *Definition for singly-linked list.
 *struct ListNode {
 *   int val;
 *   ListNode * next;
 *   ListNode() : val(0), next(nullptr) {}
 *   ListNode(int x) : val(x), next(nullptr) {}
 *   ListNode(int x, ListNode *next) : val(x), next(next) {}
 *};
 */
class Solution
{
    public:
        ListNode* reverseList(ListNode *head)
        {
            if (!head || head->next == NULL)
            {
                return head;
            }
            ListNode *prev = NULL;
            ListNode *curNode = head;
            ListNode *nextNode = head->next;
            while (curNode->next)
            {
                curNode->next = prev;
                prev = curNode;
                curNode = nextNode;
                nextNode = nextNode->next;
            }
            curNode->next = prev;
            return curNode;
        }
        ListNode* removeNodes(ListNode *head)
        {
            if (!head || !head->next)
            {
                return head;
            }
            head = reverseList(head);
            ListNode *i = head, *j = head;
            while (j)
            {
                if (i->val > j->val)
                {
                    i->next = j->next;
                    // i = j->next;
                }
                else
                {
                    i = j;
                }
                j = j->next;
            }
            head = reverseList(head);
            return head;
        }
};

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseLL(ListNode head){
        if(head == null || head.next == null)
            return head;
        
        ListNode newHead = reverseLL(head.next);
        ListNode tail = head.next;
        
        head.next = null;
        tail.next = head;
        
        return newHead;
    }
    
    public ListNode removeNodes(ListNode head) {
        if(head.next == null)
            return head;
        
        head = reverseLL(head);
        ListNode res = head;
        ListNode tail = head;
        ListNode temp = head.next;
        
        while(temp != null){
            if(temp.val >= tail.val){
                tail.next = temp;
                tail = temp;
            }
            
            temp = temp.next;
        }
        
        tail.next = null;
        
        return reverseLL(res);
    }
}